File: apps/frontend/src/lib/scan-root-warnings.ts
/** Mirror of Rust `scan_root_warnings` — keep pattern ids aligned. */

export type ScanRootWarningId =
  | 'cargo_home'
  | 'rustup_home'
  | 'npm_cache'
  | 'pnpm_store'
  | 'yarn_cache'
  | 'go_mod_cache'
  | 'go_build_cache';

export type ScanRootWarning = {
  id: ScanRootWarningId;
  path: string;
};

function normPath(path: string): string {
  return path.trim().replace(/\//g, '\\').replace(/\\+$/, '').toLowerCase();
}

type Pattern = { id: ScanRootWarningId; test: (normalized: string) => boolean };

const PATTERNS: Pattern[] = [
  {
    id: 'cargo_home',
    test: (p) => p === '.cargo' || p.endsWith('\\.cargo'),
  },
  {
    id: 'rustup_home',
    test: (p) => p === '.rustup' || p.endsWith('\\.rustup'),
  },
  {
    id: 'npm_cache',
    test: (p) => p.includes('\\appdata\\local\\npm-cache'),
  },
  {
    id: 'pnpm_store',
    test: (p) => p.includes('\\appdata\\local\\pnpm'),
  },
  {
    id: 'yarn_cache',
    test: (p) => p.includes('\\appdata\\local\\yarn'),
  },
  {
    id: 'go_mod_cache',
    test: (p) => p.includes('\\go\\pkg\\mod'),
  },
  {
    id: 'go_build_cache',
    test: (p) => p.endsWith('\\go-build'),
  },
];

export function scanRootWarning(path: string): ScanRootWarning | undefined {
  const normalized = normPath(path);
  if (!normalized) return undefined;
  for (const pattern of PATTERNS) {
    if (pattern.test(normalized)) {
      return { id: pattern.id, path: path.trim() };
    }
  }
  return undefined;
}

export function scanRootsWarnings(paths: readonly string[]): ScanRootWarning[] {
  const out: ScanRootWarning[] = [];
  const seen = new Set<string>();
  for (const p of paths) {
    const w = scanRootWarning(p);
    if (!w) continue;
    const key = `${w.id}:${w.path.toLowerCase()}`;
    if (seen.has(key)) continue;
    seen.add(key);
    out.push(w);
  }
  return out;
}

File: apps/desktop/src-tauri/src/util/scan_root_warnings.rs
//! Warn when scan roots overlap global toolchain caches (v0.9.10).

use serde::Serialize;
use std::path::Path;

#[derive(Debug, Clone, PartialEq, Eq, Serialize)]
pub struct ScanRootWarning {
    pub id: &'static str,
    pub path: String,
}

fn norm_path(path: &str) -> String {
    path.replace('/', "\\").trim().trim_end_matches('\\').to_lowercase()
}

fn path_matches_pattern(normalized: &str, pattern: &ScanRootPattern) -> bool {
    match pattern.match_kind {
        ScanRootMatchKind::EndsWith(suffix) => {
            normalized == suffix || normalized.ends_with(&format!("\\{suffix}"))
        }
        ScanRootMatchKind::Contains(part) => normalized.contains(part),
    }
}

enum ScanRootMatchKind {
    EndsWith(&'static str),
    Contains(&'static str),
}

struct ScanRootPattern {
    id: &'static str,
    match_kind: ScanRootMatchKind,
}

const PATTERNS: &[ScanRootPattern] = &[
    ScanRootPattern {
        id: "cargo_home",
        match_kind: ScanRootMatchKind::EndsWith(".cargo"),
    },
    ScanRootPattern {
        id: "rustup_home",
        match_kind: ScanRootMatchKind::EndsWith(".rustup"),
    },
    ScanRootPattern {
        id: "npm_cache",
        match_kind: ScanRootMatchKind::Contains("\\appdata\\local\\npm-cache"),
    },
    ScanRootPattern {
        id: "pnpm_store",
        match_kind: ScanRootMatchKind::Contains("\\appdata\\local\\pnpm"),
    },
    ScanRootPattern {
        id: "yarn_cache",
        match_kind: ScanRootMatchKind::Contains("\\appdata\\local\\yarn"),
    },
    ScanRootPattern {
        id: "go_mod_cache",
        match_kind: ScanRootMatchKind::Contains("\\go\\pkg\\mod"),
    },
    ScanRootPattern {
        id: "go_build_cache",
        match_kind: ScanRootMatchKind::EndsWith("\\go-build"),
    },
];

/// Returns a warning id when `path` looks like a global toolchain cache root.
pub fn scan_root_warning(path: &Path) -> Option<ScanRootWarning> {
    let display = path.to_string_lossy().to_string();
    if display.trim().is_empty() {
        return None;
    }
    let normalized = norm_path(&display);
    for pattern in PATTERNS {
        if path_matches_pattern(&normalized, pattern) {
            return Some(ScanRootWarning {
                id: pattern.id,
                path: display,
            });
        }
    }
    None
}

pub fn scan_roots_warnings(paths: &[String]) -> Vec<ScanRootWarning> {
    let mut out = Vec::new();
    let mut seen = std::collections::HashSet::new();
    for p in paths {
        if let Some(w) = scan_root_warning(Path::new(p)) {
            let key = format!("{}:{}", w.id, w.path.to_lowercase());
            if seen.insert(key) {
                out.push(w);
            }
        }
    }
    out
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn warns_cargo_home() {
        let w = scan_root_warning(Path::new(r"C:\Users\me\.cargo")).expect("warn");
        assert_eq!(w.id, "cargo_home");
    }

    #[test]
    fn warns_npm_cache() {
        let w = scan_root_warning(Path::new(
            r"C:\Users\me\AppData\Local\npm-cache",
        ))
        .expect("warn");
        assert_eq!(w.id, "npm_cache");
    }

    #[test]
    fn allows_project_root() {
        assert!(scan_root_warning(Path::new(r"E:\repos\my-app")).is_none());
    }
}
